:ship: dockerise and schedule

markysoft opened the pull request.

markysoft committed:
:ship: dockerise and schedule, drop coverage temporarily

markysoft committed:
:clipboard: improved test coverage, removed code only used in test

markysoft committed:
:eyes: move nock to dev dependencies

markysoft committed:
:lock: updated synk due to nock

markysoft committed:
:lock: synk changes again

markysoft committed:
:clipboard: fixed problem test

markysoft committed:
:boat: run tests in container, added nginx

markysoft committed:
:eyes: typo in travis.yml

markysoft committed:
:eyes: bad indentation in travis.yml

markysoft committed:
:eyes: more travis tweaks

markysoft committed:
:lock: upgraded snyk

markysoft committed:
:sweat_drops: synk fails in container

markysoft committed:
:wrench: snyk fixes, run node as nodeuser in supervisord

markysoft committed:
:file_folder: create output folder

markysoft committed:
:unlock: disabled problem snyk tests - running as other build step

c2s-dev commented:
:rocket: deployment of pomi-data-etl succeeded (http://pomi-data-etl-pr-16.dev.beta.nhschoices.net)

markysoft committed:
:book: updated readme with scheduler and urls to generated files

markysoft committed:
:eyes: typo in readme

c2s-dev commented:
:rocket: deployment of pomi-data-etl succeeded (http://pomi-data-etl-pr-16.dev.beta.nhschoices.net)

c2s-dev commented:
:rocket: deployment of pomi-data-etl succeeded (http://pomi-data-etl-pr-16.dev.beta.nhschoices.net)

markysoft committed:
:cd: add persistent volume for output files, fallback haproxy label

c2s-dev commented:
:rocket: deployment of pomi-data-etl succeeded (http://pomi-data-etl-pr-16.dev.beta.nhschoices.net)

markysoft committed:
:wrench: build changes as per gp-data-etl

c2s-dev commented:
:rocket: deployment of pomi-data-etl succeeded (http://pomi-data-etl-pr-16.dev.beta.nhschoices.net)

st3v3nhunt commented on .travis.yml:
```diff
@@ -1,6 +1,23 @@
+  - docker-compose -f docker-compose-tests.yml run travis-tests
+
+after_success:
+  - docker-compose -f docker-compose-tests.yml build travis-tests-integration
```
There is no script for `travis-tests-integration`. Suggest this and the following line are removed.

markysoft replied:
Files are not available until after the ETL has run, hence the 404 errors.
Regarding the console.time I'd rather this be raised as an issue, this work is primarily concerned with the scheduling and file publishing, no changes have been made to the underlying ETL.

Thanks, didn't notice the integration tests line.

markysoft replied:
I see, 404 errors occur even though the file is present locally. Works OK when deployed. This is caused by symlinks between docker volumes and host folders. The symlink created in docker does not resolve to the host folder, and creating one during the start script causes the docker build to fail as it's trying to create an existing link.

This has been addresses by writing directly to the html/json folder so a symlink is not required.

st3v3nhunt commented on package.json:
```diff
@@ -11,12 +11,16 @@
+    "snyk": "snyk",
+    "snyk-travis": "yarn run snyk -- test --org=nhsuk && yarn run snyk -- monitor --org=nhsuk",
+    "start": "node schedule.js",
+    "test": "snyk test && mocha --recursive",
```
I'd suggest `snyk` was not done during `test`. This would also mean the `SYNK_TOKEN` can be removed from the `test-etl` service in `docker-compose-tests.yml`

markysoft replied:
Good call.

st3v3nhunt commented on scripts/test:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+docker-compose down -v
```
The file argument (`-f docker-compose-tests.yml`) needs adding here for the `down` to be specific to the tests, AFAIK.

st3v3nhunt commented on docker-compose-tests.yml:
```diff
@@ -0,0 +1,65 @@
+      - pomi-etl-node-modules-test:/code/node_modules
+    environment:
+      NODE_ENV: test
+      SNYK_TOKEN: ${SNYK_TOKEN}
```
This could go, if you agree on not running `snyk` as part of `test`. Comment has been added against `package.json`

st3v3nhunt reviewed (dismissed):
Do you think the use of `console.time` is required given the log statements are all time stamped and individual blocks of activity could be calculated without the specific use of `console.time`. The reason I ask is it would be good to not use the `console` to log out anything.

I've run the job locally (in a container) and accessed the web page with the link to the files on and I get a `404` for all of the files.

markysoft committed:
:do_not_litter: removed unwanted code and tests
removed steps in travis.yml that are not used
removed services in docker-compose that are not used
added mising -f switch to test file
removed synk test from local test run

c2s-dev commented:
:rocket: deployment of pomi-data-etl succeeded (http://pomi-data-etl-pr-16.dev.beta.nhschoices.net)

st3v3nhunt commented:
Issue created for removal of `console.time*` https://github.com/nhsuk/pomi-data-etl/issues/17

markysoft committed:
:wrench: changed output folder to html/json to avoid symlink

c2s-dev commented:
:rocket: deployment of pomi-data-etl succeeded (http://pomi-data-etl-pr-16.dev.beta.nhschoices.net)

neilbmclaughlin commented on README.md:
```diff
@@ -18,12 +18,14 @@ the indicator reference number i.e. P02154
+
+Once initiated the scrape will download the files, strip out any records that are
+not for the current latest period (calculated based on the records), create csv
+file(s) containing those records in the output dir (`./html/json/`) and create json 
```
Didn't expect the CSV files to be in the ./html/json directory. Maybe a ./tmp directory?

markysoft replied:
I think the issue is more the intermediate files are not cleaned up after processing. The scope of this PR it to containerise, schedule and expose the files, whereas this is changes to the operation of the underlying ETL.
I'd rather this be addressed as a separate issue rather than bundle it in with this PR.
Have created issue https://github.com/nhsuk/pomi-data-etl/issues/20

neilbmclaughlin commented on README.md:
```diff
@@ -18,12 +18,14 @@ the indicator reference number i.e. P02154
-dir (`./output/`) and create json file(s) containing an array of objects in
-the form of
+Running `scripts/start` will bring up a docker container hosting a web server and initiate the scrape at a scheduled time.
+The default is 11pm. To test locally set an environment variable `ETL_SCHEDULE` to a new time, i.e. `export ETL_SCHEDULE='25 15 * * *'` to start the processing at 3:25pm.
```
The TZ for the container for seems to be GMT. Might be worth mentioning this - first time I ran it I entered the current BST time and it took a little while to figure out why the download hadn't run.

neilbmclaughlin commented on README.md:
```diff
@@ -37,4 +39,9 @@ with an `(I)` appended e.g. `EMIS (I)`. The addition of `(I)` represents a GP
-`jq -c '[.[].Supplier] | unique ' output/pomi.json`
+`jq -c '[.[].Supplier] | unique ' html/json/pomi.json`
+
+The json files created will be available via the container's webserver at
```
I can see that the `booking` file would contain the Booking system information and can guess that the `scripts` file is for Prescription request system but I'm not sure what the records download is for. I could look at the code and work it out but it would be best if the readme explicitly mentioned it.

Also, http://localhost/json/records.json currently returns 404.

markysoft replied:
It's missing a description along with Booking and scripts at lines 13-14. I'll add the missing line.

The JSON files are not available until the ETL has completed. I've added a line to the readme to make this clearer.

neilbmclaughlin replied:
ETL has been running since yesterday and still getting a '404' on records.json.

There was an error on load (need to scroll right to see it) which I thought the process had recovered from:

```
etl_1       | 2017-05-30 14:37:04,835 CRIT Supervisor running as root (no user in config file)
etl_1       | 2017-05-30 14:37:04,842 INFO supervisord started with pid 1
etl_1       | 2017-05-30 14:37:05,846 INFO spawned: 'nginx' with pid 7
etl_1       | 2017-05-30 14:37:05,849 INFO spawned: 'etl' with pid 8
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Scheduling job with rule '38 14 * * *'","time":"2017-05-30T14:37:06.026Z","v":0}
etl_1       | 2017-05-30 14:37:07,033 INFO success: nginx entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
etl_1       | 2017-05-30 14:37:07,034 INFO success: etl entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Starting download and transformation of BOOKING data","time":"2017-05-30T14:38:00.053Z","v":0}
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Starting download and transformation of SCRIPTS data","time":"2017-05-30T14:38:00.058Z","v":0}
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Starting download and transformation of RECORDS data","time":"2017-05-30T14:38:00.058Z","v":0}
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":50,"msg":"Processing failed { Error: getaddrinfo EAI_AGAIN indicators.hscic.gov.uk:443\n    at Object.exports._errnoException (util.js:1050:11)\n    at errnoException (dns.js:33:15)\n    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:73:26)\n  code: 'EAI_AGAIN',\n  errno: 'EAI_AGAIN',\n  syscall: 'getaddrinfo',\n  hostname: 'indicators.hscic.gov.uk',\n  host: 'indicators.hscic.gov.uk',\n  port: 443 }","time":"2017-05-30T14:38:05.103Z","v":0}
etl_1       | Downloading the BOOKING file took: 50493.840ms
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"html/json/booking.csv saved","time":"2017-05-30T14:38:50.890Z","v":0}
etl_1       | Downloading the SCRIPTS file took: 51005.848ms
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"html/json/scripts.csv saved","time":"2017-05-30T14:38:51.377Z","v":0}
etl_1       | Removing redundant columns from BOOKING took: 22800.550ms
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Records processed for BOOKING data column removal: 319524","time":"2017-05-30T14:39:13.692Z","v":0}
etl_1       | Getting latest period took: 0.949ms
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Latest period is 28/02/2017","time":"2017-05-30T14:39:13.693Z","v":0}
etl_1       | Removing redundant columns from SCRIPTS took: 22450.307ms
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Records processed for SCRIPTS data column removal: 320169","time":"2017-05-30T14:39:13.828Z","v":0}
etl_1       | Getting latest period took: 0.188ms
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Latest period is 28/02/2017","time":"2017-05-30T14:39:13.828Z","v":0}
etl_1       | Removing old records for BOOKING took: 5115.591ms
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Current record count: 14880","time":"2017-05-30T14:39:18.809Z","v":0}
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Old record count: 305288","time":"2017-05-30T14:39:18.809Z","v":0}
etl_1       | Download and transforming BOOKING data took: 78754.571ms
etl_1       | Removing old records for SCRIPTS took: 5055.330ms
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Current record count: 14880","time":"2017-05-30T14:39:18.884Z","v":0}
etl_1       | {"name":"pomi-etl","hostname":"94c98b7958bf","pid":8,"level":30,"msg":"Old record count: 305288","time":"2017-05-30T14:39:18.884Z","v":0}
etl_1       | Download and transforming SCRIPTS data took: 78826.673ms
etl_1       | Converting CSV to JSON for BOOKING took: 429.117ms
etl_1       | Converting CSV to JSON for SCRIPTS took: 361.305ms
```

I've rerun this and now I no longer get the '404'. I guess the 404 is a consequence of it failing the first time it runs when it does not have the previous version to serve. Not sure what the db loader does if one of the set of json documents is not available.

markysoft replied:
Summary of discussion: the missing file was caused by a network error. At the moment all three files are downloaded concurrently which is fine for 3 files, but may need to move to a sequential approach if more file are added.

Created an issue to add retries on network timeouts https://github.com/nhsuk/pomi-data-etl/issues/18

neilbmclaughlin commented on etl.js:
```diff
@@ -0,0 +1,18 @@
+    { type: 'RECORDS', url: codedRecordsDownloadUrl },
+  ];
+
+  downloads.forEach((request) => {
```
`request` is an overloaded term in an express app. Maybe `download`?

neilbmclaughlin commented on html/index.html:
```diff
@@ -0,0 +1,60 @@
+               <ul>
+                  <li><a href='json/booking.json'>booking.json</a>: array of GP surgery booking details</li>
+                  <li><a href='json/scripts.json'>scripts.json</a>: array of GP surgery prescription details</li>
+                  <li><a href='json/records.json'>records.json</a>: array of GP surgery rercords details</li>
```
typo: _rercords_

neilbmclaughlin commented on html/index.html:
```diff
@@ -0,0 +1,60 @@
+<!doctype html>
```
Not sure what the purpose of this file is. 

If we need it we probably don't need the clinical assurance and cookie links

markysoft replied:
The file is there to provide easy access to the produced JSON files. I'll remove the footer.

neilbmclaughlin commented on test/integration/downloadAndProcessFile.js:
```diff
@@ -19,17 +21,26 @@ const SUPPLIER_HEADER = constants.HEADERS.SUPPLIER;
+      .reply(200, stubbedData);
+
     before('delete files and run process', () => {
       if (fs.existsSync(`${OUTPUT_DIR}/${testFilePath}`)) {
```
fs.exists is now deprecated (https://nodejs.org/api/fs.html#fs_fs_exists_path_callback). Use fs.statSync instead?

markysoft replied:
I've used fs.accessSync instead. The behaviour of the new function is to throw an error rather than return a value so have created a utility method in fsHelper

neilbmclaughlin commented on test/integration/downloadAndProcessFile.js:
```diff
@@ -19,17 +21,26 @@ const SUPPLIER_HEADER = constants.HEADERS.SUPPLIER;
         fs.unlinkSync(`${OUTPUT_DIR}/${testFilePath}`);
       }
 
       return Promise.resolve(
```
Nice to see how to use promises in the `before`to allow setup to complete.

I'll be copying this sometime soon!

neilbmclaughlin commented on test/unit/apiRequest.js:
```diff
@@ -4,11 +4,13 @@ const apiRequest = require('../../app/lib/apiRequest');
 describe('apiRequest', () => {
-  it('should error if the status code is not 2xx', () =>
+  it('should error if the status code is not 2xx', function test() {
+    this.timeout = 10000;
```
Is it worth a comment about not using an arrow function because of the need to bind to `this`?

markysoft replied:
Good idea, I've also moved the test to integration as it is not a unit test.

neilbmclaughlin commented on app/lib/downloadAndProcessFile.js:
```diff
@@ -0,0 +1,39 @@
+  log.error('Processing failed', err);
+}
+
+function startTimer(request, timerMsg) {
```
saveFile, removeAllRecords, etc are in their own files. Should startTimer too?

markysoft replied:
Have extracted it into own file and added a test.

neilbmclaughlin commented on app/lib/constants.js:
```diff
@@ -4,5 +4,5 @@ module.exports = {
     SUPPLIER: 'Supplier',
   },
-  OUTPUT_DIR: 'output',
+  OUTPUT_DIR: 'html/json',
```
One path, two file formats. And the directory contains CSV's. Just saying.

markysoft replied:
Will only contain one file format once issue https://github.com/nhsuk/pomi-data-etl/issues/20 is addressed.

neilbmclaughlin replied:
What I mean is the directory is called 'html' when it contains css, json, csv as well as html. Would it be better named 'webroot' or something along those lines?

markysoft committed:
:eyes: refactoring, fixed typos, removed html footer

c2s-dev commented:
:rocket: deployment of pomi-data-etl succeeded (http://pomi-data-etl-pr-16.dev.beta.nhschoices.net)

markysoft committed:
:clipboard: renamed html folder to site
